File: src/fsharp/examples/tridiag_solver/Solver.fs
(*** hide ***)
module Tutorial.Fs.examples.tridiagSolver.Solver

open System
open Microsoft.FSharp.Quotations
open Alea.CUDA
open Alea.CUDA.IL
open Alea.CUDA.Utilities
open NUnit.Framework
open FsUnit

(**
# Parallel Tridiagonal Linear System Solver 

Optimized implementation of the parallel cyclic reduction solver for 
multiple systems with dimensions smaller than the maximal number of threads per block.
A single system is solved with a thread block and each thread handles a single row 
of the system.
*)

(*** define:triDiagSolverModule ***)
[<AOTCompile>]
type TriDiagSolverModule(target) =
    inherit ILGPUModule(target)

    // core solver function
    //     n  the dimension of the tridiagonal system, must fit into one block
    //     l  lower diagonal
    //     d  diagonal
    //     u  upper diagonal
    //     h  right hand side and solution at exit
    [<ReflectedDefinition>]
    let solve n (l:deviceptr<float>) (d:deviceptr<float>) (u:deviceptr<float>) (h:deviceptr<float>) =
        let rank = threadIdx.x

        let mutable ltemp = 0.0
        let mutable utemp = 0.0
        let mutable htemp = 0.0
        
        let mutable span = 1
        while span < n do
              
            if rank < n then
                if rank - span >= 0 then
                    ltemp <- if d.[rank - span] <> 0.0 then -l.[rank] / d.[rank - span] else 0.0
                else
                    ltemp <- 0.0
                if rank + span < n then
                    utemp <- if d.[rank + span] <> 0.0 then -u.[rank] / d.[rank + span] else 0.0
                else
                    utemp <- 0.0
                htemp <- h.[rank]
            
            __syncthreads()

            if rank < n then    
                if rank - span >= 0 then              
                    d.[rank] <- d.[rank] + ltemp * u.[rank - span]
                    htemp <- htemp + ltemp * h.[rank - span]
                    ltemp <-ltemp * l.[rank - span]
                
                if rank + span < n then               
                    d.[rank] <- d.[rank] + utemp * l.[rank + span]
                    htemp <- htemp + utemp * h.[rank + span]
                    utemp <- utemp * u.[rank + span]
                           
            __syncthreads()
            
            if rank < n then
                l.[rank] <- ltemp
                u.[rank] <- utemp
                h.[rank] <- htemp

            __syncthreads()

            span <- 2*span
               
        if rank < n then
            h.[rank] <- h.[rank] / d.[rank]

    [<Kernel;ReflectedDefinition>]
    member this.Kernel (n:int) (dl:deviceptr<float>) (dd:deviceptr<float>) (du:deviceptr<float>) (db:deviceptr<float>) (dx:deviceptr<float>) =
        let tid = threadIdx.x
        let gid = blockIdx.x * n + tid

        let shared = __shared__.ExternArray<float>() |> __array_to_ptr
        let l = shared
        let d = l + n
        let u = d + n
        let b = u + n
        
        l.[tid] <- dl.[gid]
        d.[tid] <- dd.[gid]
        u.[tid] <- du.[gid]
        b.[tid] <- db.[gid]
        
        __syncthreads()

        solve n l d u b

        dx.[gid] <- b.[tid]

    member this.Apply (numSystems:int, n:int, dl:deviceptr<float>, dd:deviceptr<float>, du:deviceptr<float>, db:deviceptr<float>, dx:deviceptr<float>) =
        let sharedSize = 9 * n * __sizeof<float>()
        let lp = LaunchParam(numSystems, n, sharedSize)
        this.GPULaunch <@ this.Kernel @> lp n dl dd du db dx

    member this.Solve(numSystems, l:float[], d:float[], u:float[], b:float[]) =
        let n = d.Length / numSystems

        // check resource availability
        let sharedSize = 9 * n * __sizeof<float>()
        let maxBlockDimX = this.GPUWorker.Device.Attributes.MAX_BLOCK_DIM_X
        let maxGridDimX = this.GPUWorker.Device.Attributes.MAX_GRID_DIM_X
        let maxThreads = this.GPUWorker.Device.Attributes.MAX_THREADS_PER_BLOCK
        let maxSharedSize = this.GPUWorker.Device.Attributes.MAX_SHARED_MEMORY_PER_BLOCK
        if numSystems > maxGridDimX then failwithf "numSystems(%d) > maxGridDimX(%d)" numSystems maxGridDimX  
        if n > maxBlockDimX then failwithf "n(%d) > maxBlockDimX(%d)" n maxBlockDimX  
        if n > maxThreads then failwithf "n(%d) > maxThreads(%d)" n maxThreads 
        if sharedSize > maxSharedSize then failwithf "sharedSize(%d) > maxSharedSize(%d)" sharedSize maxSharedSize  

        use dl = this.GPUWorker.Malloc(l)
        use dd = this.GPUWorker.Malloc(d)
        use du = this.GPUWorker.Malloc(u)
        use db = this.GPUWorker.Malloc(b)
        use dx = this.GPUWorker.Malloc(d.Length)

        this.Apply(numSystems, n, dl.Ptr, dd.Ptr, du.Ptr, db.Ptr, dx.Ptr)
        dx.Gather()

(**
We test the tridiagonal solver by setting up multiple diagonally dominant sytems.
*)

(*** define:triDiagSolverTest ***)
[<Test>]
let triDiagSolverTest () =
    let rng = System.Random(42)
    
    let random a b _ =
        rng.NextDouble() * (b - a) + a
    
    let multiply (l:float[]) (d:float[]) (u:float[]) (x:float[]) =
        let n = d.Length
        let b = Array.zeroCreate n
        b.[0] <- d.[0] * x.[0] + u.[0] * x.[1]
        for i = 1 to n - 2 do
            b.[i] <- l.[i] * x.[i - 1] + d.[i] * x.[i] + u.[i] * x.[i + 1]
        b.[n - 1] <- l.[n - 1] * x.[n - 2] + d.[n - 1] * x.[n - 1]
        b
    
    let generateDiagonallyDominantSystem n =
        let l = Array.init n (random -100.0 100.0)
        let u = Array.init n (random -100.0 100.0)
        let d = Array.map2 (fun l u -> random  0.0 10.0 () + (abs l) + (abs u)) l u
        let expectedSolution = Array.init n (random -20.0 20.0)
        let b = multiply l d u expectedSolution
        expectedSolution, l, d, u, b

    let unzip5 = Array.fold (fun (acc1, acc2, acc3, acc4, acc5) (e1, e2, e3, e4, e5) -> (Array.append acc1 e1, Array.append acc2 e2, Array.append acc3 e3, Array.append acc4 e4, Array.append acc5 e5)) ([||],[||],[||],[||],[||])
    
    let n = 100
    let expectedSolutions, l, d, u, b = Array.init n (fun _ -> generateDiagonallyDominantSystem 512) |> unzip5
    use tridiagSolver = new TriDiagSolverModule(GPUModuleTarget.DefaultWorker)
    let solutions = tridiagSolver.Solve(n, l, d, u, b)
    solutions |> should (equalWithin 1e-12) expectedSolutions

File: src/csharp/examples/tridiag_solver/Solver.cs
using System;
using System.Linq;
using Alea.CUDA;
using Alea.CUDA.IL;
using NUnit.Framework;

namespace Tutorial.Cs.examples.tridiagSolver
{
    //[triDiagSolverModule]
    [AOTCompile]
    internal class TriDiagSolverModule : ILGPUModule
    {
        public TriDiagSolverModule(GPUModuleTarget target)
            : base(target)
        {
        }

        // core solver function
        //     n  the dimension of the tridiagonal system, must fit into one block
        //     l  lower diagonal
        //     d  diagonal
        //     u  upper diagonal
        //     h  right hand side and solution at exit
        public static void Solve(int n, deviceptr<double> l, deviceptr<double> d, deviceptr<double> u, deviceptr<double> h)
        {
            var rank = threadIdx.x;

            var ltemp = 0.0;
            var utemp = 0.0;
            var htemp = 0.0;

            var span = 1;
            while (span < n)
            {
                if (rank < n)
                {
                    ltemp = (rank - span >= 0) ?
                        (d[rank - span] != 0.0) ? -l[rank] / d[rank - span] : 0.0
                        :
                        0.0;


                    utemp = (rank + span < n) ?
                        (d[rank + span] != 0.0) ? -u[rank] / d[rank + span] : 0.0
                        :
                        0.0;

                    htemp = h[rank];
                }

                Intrinsic.__syncthreads();

                if (rank < n)
                {
                    if (rank - span >= 0)
                    {
                        d[rank] = d[rank] + ltemp * u[rank - span];
                        htemp = htemp + ltemp * h[rank - span];
                        ltemp = ltemp * l[rank - span];
                    }
                    if (rank + span < n)
                    {
                        d[rank] = d[rank] + utemp * l[rank + span];
                        htemp = htemp + utemp * h[rank + span];
                        utemp = utemp * u[rank + span];
                    }
                }

                Intrinsic.__syncthreads();

                if (rank < n)
                {
                    l[rank] = ltemp;
                    u[rank] = utemp;
                    h[rank] = htemp;
                }

                Intrinsic.__syncthreads();

                span *= 2;
            }

            if (rank < n) h[rank] = h[rank] / d[rank];
        }

        // multiple systems solved in different thread blocks
        [Kernel]
        public void Kernel(int n, deviceptr<double> dl, deviceptr<double> dd, deviceptr<double> du, deviceptr<double> db, deviceptr<double> dx)
        {
            var tid = threadIdx.x;
            var gid = blockIdx.x * n + tid;

            var shared = Intrinsic.__array_to_ptr(__shared__.ExternArray<double>());
            var l = shared;
            var d = l + n;
            var u = d + n;
            var b = u + n;

            l[tid] = dl[gid];
            d[tid] = dd[gid];
            u[tid] = du[gid];
            b[tid] = db[gid];

            Intrinsic.__syncthreads();

            Solve(n, l, d, u, b);

            dx[gid] = b[tid];
        }

        public void Apply(int numSystems, int n, deviceptr<double> dl, deviceptr<double> dd, deviceptr<double> du,
            deviceptr<double> db, deviceptr<double> dx)
        {
            var sharedSize = 9*n*sizeof (double);
            var lp = new LaunchParam(numSystems, n, sharedSize);
            this.GPULaunch(this.Kernel, lp, n, dl, dd, du, db, dx);
        }

        public double[] Apply(int numSystems, double[] l, double[] d, double[] u, double[] b)
        {
            var n = d.Length/numSystems;

            // check resource availability
            var sharedSize = 9*n*sizeof (double);
            var maxBlockDimX = this.GPUWorker.Device.Attributes.MAX_BLOCK_DIM_X;
            var maxGridDimX = this.GPUWorker.Device.Attributes.MAX_GRID_DIM_X;
            var maxThreads = this.GPUWorker.Device.Attributes.MAX_THREADS_PER_BLOCK;
            var maxSharedSize = this.GPUWorker.Device.Attributes.MAX_SHARED_MEMORY_PER_BLOCK;
            if (numSystems > maxGridDimX) throw new Exception(String.Format("numSystems({0}) > maxGridDimX({1})", numSystems, maxGridDimX));
            if (n > maxBlockDimX) throw new Exception(String.Format("n({0}) > maxBlockDimX({1})", n, maxBlockDimX));
            if (n > maxThreads) throw new Exception(String.Format("n({0}) > maxThreads({1})", n, maxThreads));
            if (sharedSize > maxSharedSize) throw new Exception(String.Format("sharedSize({0}) > maxSharedSize({1})", sharedSize, maxSharedSize));  

            using (var dl = this.GPUWorker.Malloc(l))
            using (var dd = this.GPUWorker.Malloc(d))
            using (var du = this.GPUWorker.Malloc(u))
            using (var db = this.GPUWorker.Malloc(b))
            using (var dx = this.GPUWorker.Malloc<double>(d.Length))
            {
                this.Apply(numSystems, n, dl.Ptr, dd.Ptr, du.Ptr, db.Ptr, dx.Ptr);
                return dx.Gather();
            }
        }
    }
    //[/triDiagSolverModule]

    //[triDiagSolverTest]
    public class Test
    {
        public static Random Rng = new Random(42);

        private static double random(double a, double b)
        {
            return Rng.NextDouble() * (b - a) + a;
        }

        private static double[] multiply(double[] l, double[] d, double[] u, double[] x)
        {
            var n = d.Length;
            var b = new double[n];
            b[0] = d[0] * x[0] + u[0] * x[1];
            for (var i = 1; i <= n - 2; i++)
                b[i] = l[i] * x[i - 1] + d[i] * x[i] + u[i] * x[i + 1];
            b[n - 1] = l[n - 1] * x[n - 2] + d[n - 1] * x[n - 1];
            return b;
        }
        
        private static Tuple<double[], double[], double[], double[], double[]> diagonallyDominantSystem(int n)
        {
            var l = new double[n];
            var u = new double[n];
            var d = new double[n];
            var expectedSolution = new double[n];
            for (var i = 0; i < n; i++)
            {
                l[i] = random(-100.0, 100.0);
                u[i] = random(-100.0, 100.0);
                d[i] = random(0.0, 10.0) + Math.Abs(l[i]) + Math.Abs(u[i]);
                expectedSolution[i] = random(-20.0, 20.0);
            }
            var b = multiply(l, d, u, expectedSolution);
            return new Tuple<double[], double[], double[], double[], double[]>(l, d, u, b, expectedSolution);
        }
        
        [Test]
        public static void TriDiagSolverTest()
        {
            const int numSystems = 512;

            var numbers =
                Enumerable
                .Repeat(numSystems, numSystems)
                .Select(diagonallyDominantSystem)
                .ToArray();

            var l = numbers.SelectMany(s => s.Item1).ToArray();
            var d = numbers.SelectMany(s => s.Item2).ToArray();
            var u = numbers.SelectMany(s => s.Item3).ToArray();
            var b = numbers.SelectMany(s => s.Item4).ToArray();
            var expectedSolution = numbers.SelectMany(s => s.Item5).ToArray();

            var triDiagSolver = new TriDiagSolverModule(GPUModuleTarget.DefaultWorker);

            var x = triDiagSolver.Apply(numSystems, l, d, u, b);
            var errs = new double[x.Length];
            for (var i = 0; i < x.Length; ++i)
                errs[i] = Math.Abs(expectedSolution[i] - x[i]);
            var err = errs.Max();
            Console.WriteLine("error = {0}", err);
            Assert.LessOrEqual(err, 1e-8);
        }
    }
    //[/triDiagSolverTest]
}